filters/builtin/dynamicbackendfilter.go

```diff
@@ -0,0 +1,157 @@
+// `SetOutgoingHost()` is used to set the host header
+func NewSetDynamicBackendHost() filters.Spec {
+	return &dynamicBackendFilter{typ: setDynamicBackendHost}
+}
+
+// Returns a filter specification that is used to set dynamic backend scheme.
+// Instances expect one parameters: a scheme name.
+// Name: "setDynamicBackendScheme".
+//
+// The value is put into the `StateBag`
+func NewSetDynamicBackendScheme() filters.Spec {
+	return &dynamicBackendFilter{typ: setDynamicBackendScheme}
+}
+
+// Returns a filter specification that is used to set dynamic backend url.
+// Instances expect one parameters: a url.
+// Name: "setDynamicBackendUrl".
+//
+// The value is put into the `StateBag`, additionally `SetOutgoingHost()`
+// is used to set the host header if the input provided is a valid url
+func NewSetDynamicBackendUrl() filters.Spec {
+	return &dynamicBackendFilter{typ: setDynamicBackendUrl}
+}
+
+func (spec *dynamicBackendFilter) Name() string {
+	switch spec.typ {
+	case setDynamicBackendHostFromHeader:
+		return SetDynamicBackendHostFromHeader
+	case setDynamicBackendSchemeFromHeader:
+		return SetDynamicBackendSchemeFromHeader
+	case setDynamicBackendUrlFromHeader:
+		return SetDynamicBackendUrlFromHeader
+	case setDynamicBackendHost:
+		return SetDynamicBackendHost
+	case setDynamicBackendScheme:
+		return SetDynamicBackendScheme
+	case setDynamicBackendUrl:
+		return SetDynamicBackendUrl
+	default:
+		panic("invalid type")
+	}
+}
+
+//lint:ignore ST1016 "spec" makes sense here and we reuse the type for the filter
+func (spec *dynamicBackendFilter) CreateFilter(config []interface{}) (filters.Filter, error) {
+	input, err := dynamicBackendFilterConfig(config)
+	return &dynamicBackendFilter{typ: spec.typ, input: input}, err
+}
+
+func (f *dynamicBackendFilter) Request(ctx filters.FilterContext) {
+	switch f.typ {
+	case setDynamicBackendHostFromHeader:
+		header := ctx.Request().Header.Get(f.input)
+		if header != "" {
+			ctx.StateBag()[filters.DynamicBackendHostKey] = header
+			ctx.SetOutgoingHost(header)
+		}
+	case setDynamicBackendSchemeFromHeader:
+		header := ctx.Request().Header.Get(f.input)
+		if header != "" {
+			ctx.StateBag()[filters.DynamicBackendSchemeKey] = header
+		}
+	case setDynamicBackendUrlFromHeader:
+		header := ctx.Request().Header.Get(f.input)
+		if header != "" {
+			ctx.StateBag()[filters.DynamicBackendURLKey] = header
+			bu, err := url.ParseRequestURI(header)
+			if err == nil {
+				ctx.SetOutgoingHost(bu.Host)
+			}
+		}
+	case setDynamicBackendHost:
+		ctx.StateBag()[filters.DynamicBackendHostKey] = f.input
+		ctx.SetOutgoingHost(f.input)
+	case setDynamicBackendScheme:
+		ctx.StateBag()[filters.DynamicBackendSchemeKey] = f.input
+	case setDynamicBackendUrl:
+		ctx.StateBag()[filters.DynamicBackendURLKey] = f.input
+		bu, err := url.ParseRequestURI(f.input)
+		if err == nil {
+			ctx.SetOutgoingHost(bu.Host)
+		}
+	}
+}
+
+func (f *dynamicBackendFilter) Response(ctx filters.FilterContext) {}
```

Comment:
you don't need the names only the types in an empty function

Reply:
Do we need to define so many state bag keys? I would say that we should have only one key for setting  Backend Url and all other filters should just set this value accordingly. 

To set scheme, all you need to do is take current backend url, change the scheme and set the new value back to backend url. 

To take backend url form host, you read its value and set the value to Backend url key.